CleverCloud/clever-components
# feat(forms): add framework for `ElementInternals` support

## What is this PR?

This PR adds support for `ElementInternals`. It makes our inputs (`cc-input-text`, `cc-input-number`, `cc-input-date`, `cc-select`) and our button (`cc-button`) implement the [ElementInternals](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals) API.

With this PR you get the following improvements:
- one can use our inputs inside a `<form>`, they will behave like native HTML inputs.
- our inputs have
  - a `validationMessage` property that will be displayed below the input (and not as a native tooltip)
  - a `customValidator` property that lets developers add their validation logic
  - a `customErrorMessages` property that lets developers add their error messages and also, override default error messages.

Also, it adds some utility codes that will make our lives better when dealing with forms inside high-level components:
- a `<form>` submit event handler that handles all the logic on form submission (validation, error message display, focus)
- a Lit directive to be attached to a `<form>` element
- a Lit reactive controller when dealing with asynchronous code (typically, when submitting forms that perform HTTP call)

Finally, this PR contains modifications on 3 high-level components containing a form (`cc-email-list`, `cc-orga-member-list`, `cc-ssh-key-list`). It implements form management using this new framework.

## How to review?

After many in-sync meetings, you should have a good understanding of the key concepts manipulated in this PR.

Now,
- take a look at the code,
- try the demos in sandbox (and read the code of the demo at the same time),
- check the Storybook documentation and stories,
- if you want to go further, you could take a component that has got a form, and try to use the framework.
